DI rewrite (#19271)
dotnet/AspNetCore.Docs: aspnetcore/fundamentals/dependency-injection.md

[!code-csharp[](dependency-injection/samples/3.x/DependencyInjectionSample/Services/MyDependency.cs?name=snippet1)]

> In the sample app, the `IMyDependency` service is registered with the concrete type `MyDependency`. The registration service lifetime is <xref:Microsoft.Extensions.DependencyInjection.ServiceCollectionServiceExtensions.AddScoped*>, the lifetime of a single request. [Service lifetimes](#service-lifetimes) are described later in this topic.

Reviewer: @Rick-Anderson Curious as to why this is AddScoped?
Since the dependency was explicitly created in the example above, feels it is more to be a Transient (Just from the context of this example). I might be missing something here.

Author: We could change it to transient but I don't see much advantage. The sample is used to explain how to use dependency injection. It's not a good sample because as we show later, the logger is already in the DI container. I'd like to have a simple sample to demonstrate DI  where the code required DI.

@serpent5  should we change the sample to transient?

The container resolves `ILogger<TCategoryName>` by taking advantage of [(generic) open types](/dotnet/csharp/language-reference/language-specification/types#open-and-closed-types), eliminating the need to register every [(generic) constructed type](/dotnet/csharp/language-reference/language-specification/types#constructed-types):

```csharp
// This code provided by the framework.

> services.AddSingleton(typeof(ILogger<>), typeof(Logger<>));

Reviewer 2: Maybe we should remove this as an example since it's never relevant in practice

Author: Will do.
BTW, this is not ready for review but comments like that are welcome.

* When `Startup.ConfigureServices` is run and an instance is specified with the service registration.

Every subsequent request uses the same instance. If the app requires singleton behavior, allowing the service container to manage the service's lifetime is recommended. Don't implement the singleton design pattern and provide code to dispose of the singleton. Services should never be disposed by code that resolved the service from a container. If a type or factory is registered as a singleton, the container will dispose the singleton automatically.

Singleton services must be thread safe and are often used in stateless services.

In apps that process requests, singleton services are disposed when the <xref:Microsoft.Extensions.DependencyInjection.ServiceProvider> is disposed on application shutdown. Because memory is not released until the app is shut down, memory use with a singleton must be considered.

> [!WARNING]
> Do ***not*** resolve a scoped service from a singleton. It may cause the service to have incorrect state when processing subsequent requests. It's fine to resolve a singleton service from a scoped or transient service.

## Service registration methods

> Service registration extension methods offer overloads that are useful in specific scenarios.

Reviewer: Should it be mentioned here that IServiceCollection is a collection of ServiceDescriptor and an example of adding it like that explicitly (as below)? And state the extension methods makes it easier to add registrations.
`
var item = new ServiceDescriptor(typeof(IMyDependency), context = >new MyDependency(),   ServiceLifetime.Transient);

services.Add(item);
`

Author: @rahulpnath  great suggestion. Check the updated doc build.

Reviewer: lgtm :)

For more information, see <xref:fundamentals/middleware/write#per-request-middleware-dependencies>.

Do ***not*** resolve a scoped service from a singleton. It may cause the service to have incorrect state when processing subsequent requests. It's fine to:

* Resolve a singleton service from a scoped or transient service.
* Resolve a scoped service from another scoped or transient service.

By default, in the development environment, resolving a service from another service with longer lifetime throws an exception. For more information, see [Scope validation](#sv).

### Singleton

Singleton lifetime services (<xref:Microsoft.Extensions.DependencyInjection.ServiceCollectionServiceExtensions.AddSingleton*>) are created either:

* The first time they're requested.
* When `Startup.ConfigureServices` is run and an instance is specified with the service registration.

> Every subsequent request uses the same instance. If the app requires singleton behavior, allowing the service container to manage the service's lifetime is recommended. Don't implement the singleton design pattern and provide code to dispose of the singleton. Services should never be disposed by code that resolved the service from a container. If a type or factory is registered as a singleton, the container will dispose the singleton automatically.

Reviewer 3: Option 1:

```suggestion
Every subsequent request uses the same instance. If the app requires singleton behavior, allow the service container to manage the service's lifetime. Don't implement the singleton design pattern and provide code to dispose of the singleton. Services should never be disposed by code that resolved the service from a container. If a type or factory is registered as a singleton, the container will dispose the singleton automatically.
```

> Every subsequent request uses the same instance. If the app requires singleton behavior, allowing the service container to manage the service's lifetime is recommended. Don't implement the singleton design pattern and provide code to dispose of the singleton. Services should never be disposed by code that resolved the service from a container. If a type or factory is registered as a singleton, the container will dispose the singleton automatically.

Reviewer 3: Option 2:

```suggestion
Every subsequent request uses the same instance. If the app requires singleton behavior, allow the service container to manage the service's lifetime. Don't implement the singleton design pattern to manage the lifetime of the service. Services should never be disposed by code that resolved the service from a container. If a type or factory is registered as a singleton, the container will dispose the singleton automatically.
```

This is a separate suggestion because the _user code_ thing has been discussed previously and I don't want to bundle it all together. It also includes the preceding suggestion.

Singleton services must be thread safe and are often used in stateless services.

In apps that process requests, singleton services are disposed when the <xref:Microsoft.Extensions.DependencyInjection.ServiceProvider> is disposed on application shutdown. Because memory is not released until the app is shut down, memory use with a singleton must be considered.

> [!WARNING]
> Do ***not*** resolve a scoped service from a singleton. It may cause the service to have incorrect state when processing subsequent requests. It's fine to resolve a singleton service from a scoped or transient service.

## Service registration methods

`IServiceCollection` is a collection of <xref:Microsoft.Extensions.DependencyInjection.ServiceDescriptor>. The following code shows how to add a service with a constructor:

[!code-csharp[](dependency-injection/samples/3.x/DependencyInjectionSample/Startup5.cs?name=snippet)]

> 

Reviewer 3: I don't think the section should lead with this, because I expect it's rare that this approach is used. It's useful to know this is how it works behind the scenes, but I think I would:

- Move it to the end of the _Service registration methods_ section and highlight that the `AddXXX` methods are wrappers around this approach.
- Perhaps even just show an example of how e.g. `AddTransient` works, which demonstrates concept without showing it in app code.

## Service lifetimes

Choose an appropriate lifetime for each registered service. ASP.NET Core services can be configured with the following lifetimes:

### Transient

Transient lifetime services (<xref:Microsoft.Extensions.DependencyInjection.ServiceCollectionServiceExtensions.AddTransient*>) are created each time they're requested from the service container. This lifetime works best for lightweight, stateless services. Transient services don't need to be thread safe.

In apps that process requests, transient services are disposed at the end of the request.

### Scoped

Scoped lifetime services (<xref:Microsoft.Extensions.DependencyInjection.ServiceCollectionServiceExtensions.AddScoped*>) are created once per client request (connection).

When using Entity Framework Core, the [AddDbContext](/dotnet/api/microsoft.extensions.dependencyinjection.entityframeworkservicecollectionextensions.adddbcontext) extension method registers `DbContext` types with a scoped lifetime by default.

In apps that process requests, scoped services are disposed at the end of the request.

> Use scoped services in a middleware with one of the following approaches:

Author: @rahulpnath please review this change. I'm rewriting the entire doc and should be finished in a couple days. WHen I'm done, it would be great if you could review the updates. I can post a link to the new doc build.
I dropped the warning because this is not a trap you can fall into. The runtime throws and exception if you try to do this.

Reviewer: Looks good and better now!

### Singleton

Singleton lifetime services (<xref:Microsoft.Extensions.DependencyInjection.ServiceCollectionServiceExtensions.AddSingleton*>) are created either:

* The first time they're requested.
* When `Startup.ConfigureServices` is run and an instance is specified with the service registration.
<!-- REVIEW what is user code? -->

> Every subsequent request uses the same instance. If the app requires singleton behavior, allowing the service container to manage the service's lifetime is recommended. Don't implement the singleton design pattern and provide user code to manage the object's lifetime in the class.

Reviewer 3: I think _user code_ here means implementing the Singleton pattern manually. e.g. using one of [Jon Skeet's patterns](https://csharpindepth.com/articles/singleton).

Reviewer 4: I think "user code" in this case is referring code that might incorrectly manually dispose a resolved singleton.

Services should never be disposed by any code that resolved the service from a container. If a type or factory is registered as a singleton, the container will dispose the singleton for you.

If an object instance is registered as a singleton, the code that created that instance is responsible for the lifetime, but the instance still must outlive the container. A reference to the singleton instance should be kept outside of the container. The singleton should then be disposed via this external reference after the container/host has already been disposed.

* <xref:Microsoft.Extensions.DependencyInjection.Extensions.ServiceCollectionDescriptorExtensions.TryAddScoped*>
* <xref:Microsoft.Extensions.DependencyInjection.Extensions.ServiceCollectionDescriptorExtensions.TryAddSingleton*>

> [TryAddEnumerable(ServiceDescriptor)](xref:Microsoft.Extensions.DependencyInjection.Extensions.ServiceCollectionDescriptorExtensions.TryAddEnumerable*) methods register the service if there isn't already an implementation *of the same type*. Multiple services are resolved via `IEnumerable<{SERVICE}>`. When registering services, the developer should add an instance if one of the same type hasn't already been added. Generally, `TryAddEnumerable` is used by library authors to avoid registering two copies of an instance in the container.

Reviewer 3: ```suggestion
[TryAddEnumerable(ServiceDescriptor)](xref:Microsoft.Extensions.DependencyInjection.Extensions.ServiceCollectionDescriptorExtensions.TryAddEnumerable*) methods register the service if there isn't already an implementation *of the same type*. Multiple services are resolved via `IEnumerable<{SERVICE}>`. When registering services, the developer should add an instance if one of the same type hasn't already been added. Generally, library authors use `TryAddEnumerable` to avoid registering multiple copies of an implementation in the container.
```

services.TryAddEnumerable(ServiceDescriptor.Singleton<IMyDep1, MyDep>());
services.TryAddEnumerable(ServiceDescriptor.Singleton<IMyDep2, MyDep>());
services.TryAddEnumerable(ServiceDescriptor.Singleton<IMyDep1, MyDep>());
```

> Service registration is generally not order dependent except when registering multiple implementations of an interface.

Reviewer 3: ```suggestion
Service registration is generally order independent except when registering multiple implementations of the same type.
```

### Constructor injection behavior

Services can be resolved by two mechanisms:

* <xref:System.IServiceProvider>
* <xref:Microsoft.Extensions.DependencyInjection.ActivatorUtilities>:

>   * Permits object creation without service registration in the dependency injection container.

Reviewer 3: ```suggestion
  * Creates objects without service registration in the dependency injection container.
```

>   <!-- Review: what is user-facing abstractions? Frame work features or library features? -->

Reviewer 3: It looks like this means framework features.

e.g. A controller is an abstraction provided by the framework that users work with closely. Users don't usually register their controllers with DI, but those controllers can have dependencies that come from DI.

## Lifetime and registration options

> To demonstrate the difference between the lifetime and registration options, consider the following interfaces that represent tasks as an operation with an identifier, `OperationId`. Depending on how the lifetime of an operations service is configured for the following interfaces, the container provides either the same or a different instance of the service when requested by a class:

Reviewer 3: ```suggestion
To demonstrate the difference between the lifetime and registration options, consider the following interfaces that represent tasks as an operation with an identifier, `OperationId`. Depending on how the lifetime of an operation's service is configured for the following interfaces, the container provides either the same or a different instance of the service when requested by a class:
```

The sample app demonstrates object lifetimes within and between requests. The sample app's `IndexModel` and middleware requests each kind of `IOperation` type and logs the `OperationId`:

[!code-csharp[](dependency-injection/samples/3.x/DependencyInjectionSample/Pages/Index.cshtml.cs?name=snippet1)]

The middleware is similar to the `IndexModel` and resolves the same services:

[!code-csharp[](dependency-injection/samples/3.x/DependencyInjectionSample/Middleware/MyMiddleware.cs?name=snippet)]

Scoped services must be resolved in the `InvokeAsync` method:

[!code-csharp[](dependency-injection/samples/3.x/DependencyInjectionSample/Middleware/MyMiddleware.cs?name=snippet2&highlight=2)]

The logger output shows:

* *Transient* objects are always different. The transient `OperationId` value is different in the `IndexModel` and the middleware.
* *Scoped* objects are the same in each request but different across each request.
* *Singleton* objects are the same for every request.

To reduce the logging output, set "Logging:LogLevel:Microsoft:Error" in the *appsettings.Development.json* file:

> [!code-json[](dependency-injection/samples/3.x/DependencyInjectionSample/appsettings.Development.json)]

Reviewer 3: ```suggestion
[!code-json[](dependency-injection/samples/3.x/DependencyInjectionSample/appsettings.Development.json&highlight=7)]
```

* Avoid direct instantiation of dependent classes within services. Direct instantiation couples the code to a particular implementation.
* Make app classes small, well-factored, and easily tested.

If a class seems to have too many injected dependencies, that's generally a sign that the class has too many responsibilities and is violating the [Single Responsibility Principle (SRP)](/dotnet/standard/modern-web-apps-azure-architecture/architectural-principles#single-responsibility). Attempt to refactor the class by moving some of its responsibilities into a new class. Keep in mind that Razor Pages page model classes and MVC controller classes should focus on UI concerns.

### Disposal of services

> The container calls <xref:System.IDisposable.Dispose*> for the <xref:System.IDisposable> types it creates. Services should never be disposed by any code that resolved the service from a container. If a type or factory is registered as a singleton, the container will dispose the singleton.

Reviewer 3: ```suggestion
The container calls <xref:System.IDisposable.Dispose*> for the <xref:System.IDisposable> types it creates. Services should never be disposed by any code that resolved the service from a container. If a type or factory is registered as a singleton, the container disposes the singleton.
```

If a class seems to have too many injected dependencies, this is generally a sign that the class has too many responsibilities and is violating the [Single Responsibility Principle (SRP)](/dotnet/standard/modern-web-apps-azure-architecture/architectural-principles#single-responsibility). Attempt to refactor the class by moving some of its responsibilities into a new class. Keep in mind that Razor Pages page model classes and MVC controller classes should focus on UI concerns. Business rules and data access implementation details should be kept in classes appropriate to these [separate concerns](/dotnet/standard/modern-web-apps-azure-architecture/architectural-principles#separation-of-concerns).

### Disposal of services

> <!-- REVIEW: define user code.  -->

Reviewer 3: I think this means something like this:

```csharp
var mySingleton = new MySingleton();
services.AddSingleton(mySingleton);
```
`mySingleton` isn't disposed by the container because _we_ provided the instance directly.

However, these two approaches result in automatic disposal:

```csharp
services.AddSingleton<MySingleton>();
services.AddSingleton(() => new MySingleton()); // Factory.
```

This one's more interesting:

```csharp
var mySingleton = new MySingleton();
services.AddSingleton(() => mySingleton);
```

This also causes automatic disposal, because the container assumes you created an instance on its behalf. I don't think you should ever really do this, though. It'd be dangerous if you tried to use `mySingleton` after the container was disposed.

**Scenario**

The app requires an <xref:System.IDisposable> instance with a transient lifetime for either of the following scenarios:

> <!-- Review need to define root scope. Is that the same as the root container that's only disposed when the app shuts down?

Reviewer 3: This is how I understand it.
